envoy/api/v2/cds.proto

```diff
@@ -166,6 +166,12 @@ message Cluster {
   // parameter to 1 will effectively disable keep alive.
   google.protobuf.UInt32Value max_requests_per_connection = 9;
 
+  // [#not-implemented-hide:]
+  // The idle timeout for upstream connections. The idle timeout is defined as the period in which
+  // there are no active requests. If not set, there is no idle timeout. When the idle timeout is
```

Comment:
Given that this is on the cluster level, I don't think we should phrase this in terms of requests. It probably needs to be defined in terms of either bytes sent or received. The other option since this effectively HTTP specific would be to add a generic HTTP option message (we already have HTTP/1 and HTTP/2) and add something specific to connection pool connections there. Thoughts? I think I actually prefer the latter in this case.

Reply:
@mattklein123 agree on the requests phrasing. I was actually experimenting with it on the Envoy side and you are right, requests does not really makes sense here. We should not hook this timeout calculation to `newStream` and `onStreamDestroy`. Would should definitely be hooking it to bytes send/received to determine the idle. I could not find such call backs in ConnectionPool. Could you please point me to those callbacks so that i can try hooking idleness computation there?
on the placement of this, I think it may be more clearer if we keep it along side of max_requests but not very strong about it instead of introducing a new message just for this. However i am not opposed to doing it that way if you feel that is cleaner. let me know.

Reply:
See https://github.com/envoyproxy/envoy/blob/master/include/envoy/network/connection.h#L101 for finding out when bytes are sent. I haven't looked at how you'd pipe that through to the ConnectionPool.

Reply:
@mattklein123 I looked at it regarding the placement of this option. I see your point. I think introducing a new Http specific option message and adding this there would make more sense. I guess the idle time outs for Tcp also can be moved to https://github.com/envoyproxy/data-plane-api/blob/master/envoy/api/v2/core/protocol.proto#L17 to be consistent. But that could be  a different PR.
@ggreenway if you are fine with the above, I will make the change. Let me know.

Reply:
@ggreenway Thanks for the bytes sent call back pointer. I briefly looked at it today. I think we can pipe that at https://github.com/envoyproxy/envoy/blob/master/source/common/http/http2/conn_pool.cc#L226 some thing similar to
data.connection_->addBytesSentCallback()

Reply:
Yeah I agree with @ggreenway that we can likely consolidate handling of certain options like connect timeout, etc., but I think that is out of scope for this. @ramaraochavali I think my recommendation is to add an HTTP specific set of options, and actually just stick with the requests terminology and not the bytes sent terminology. (Basically what you had, but make it HTTP specific). Then it's relatively easy to implement that option inside of both the h1 and h2 connection pool.

Reply:
@mattklein123 ok. So I may have misunderstood what you said. To be clear, I will add a new http options and add this configuration. On the the Envoy side, hook this to stream create and destroy events?Another question though, how will this work for long running streams like our xDs? Maybe I am missing something here